src/Symfony/Component/Mailer/Transport/Smtp/Auth/XOAuth2Authenticator.php

```diff
@@ -32,6 +39,7 @@ public function getAuthKeyword(): string
      */
     public function authenticate(EsmtpTransport $client): void
     {
-        $client->executeCommand('AUTH XOAUTH2 '.base64_encode('user='.$client->getUsername()."\1auth=Bearer ".$client->getPassword()."\1\1")."\r\n", [235]);
+        $token = $this->tokenProvider ? $this->tokenProvider->getToken() : $client->getPassword();
```

Comment:
```suggestion
        $token = $this->tokenProvider?->getToken() : $client->getPassword();
```

But I'm wondering about this line as now we give two ways to pass that secret - what if both are set for example? I feel like it'd be better to provide only one way to pass that secret - the existing one if possible :) Does that make sense? (it doesn't have to :) )

Reply:
the issue with the existing one is that oauth tokens tend to be short-lived. In your worker process, you cannot just set the token as the password in the Transport as this will not properly refresh it when it expires.

Reply:
changing the client to fetch the password dynamically would be a large refactor, and would mix responsibilities for the auth method.

the thing is that an oauth token can not be statically configured. and in a long running process it might expire before the process ends, so each time we need it we need to be ready to re-fetch it.

we could inject some TokenProviderInterface into the client and use that for getPassword(), but i don't think that would be better.